### DISCUSSION TITLE
Framework for generic exposure of device configuration attributes

### DISCUSSION THREAD

**Discussion by dumpfheimer** [zigpy/zigpy#934] [ZHA development]
Hi!

I developed a proof of concept for simple exposure of device configuration attributes [here](https://github.com/dumpfheimer/core/tree/dumpfheimer/zha-expose-configuration-entities) .
What I mean with that is for example the start up behaviour of lights.

My development is based on @puddly 's ZCLv7 development.

It exposes integer / enum type configurations as entities, that are configured in the cluster as follows:
![image](https://user-images.githubusercontent.com/60989727/158354055-71fe48e4-a65e-4087-a0de-62cb5f698b5c.png)
![image](https://user-images.githubusercontent.com/60989727/158354166-33cd30aa-dd99-47f0-8ee0-01d0fd52e440.png)

and results in this:
![image](https://user-images.githubusercontent.com/60989727/158354322-c32b0a74-3c3e-49b6-934d-8b588018a413.png)
![image](https://user-images.githubusercontent.com/60989727/158354364-ba9d77bb-d75f-4311-b651-613c38716e2d.png)

I did this by extending the discovery of entities in zha integration and having integer/enum entities being created dynamically
https://github.com/dumpfheimer/core/blob/70816c9703c9b475a4ac4b50ad14788d1fb72096/homeassistant/components/zha/core/discovery.py#L131

This approach should be fully backwards compatible and enable zigpy and zha quirks to expose key configuration attributes very easily.

Feedback is very much appreciated.

**Comment by MattWestb** [comment]
That looks great only that some manufacturers is adding different functions in there devices or using other attribute with not exact the same function we have adding them then we have finding out how there is working.

Firs priority is ZCL R7/8 and then i thing adding the no standard can coming later.

I keep collecting info and trying documenting it for the future here: https://github.com/zigpy/zigpy/discussions/823#discussioncomment-1539469
Also updating if the attributes is used in some quirks or if its moved to tuya INIT as one "standard" tuya attribute so being  easier using them in our system.

Great work and hope its can coming in the future in ZHA also startup light level and color (temperature) for lights.

**Reply by dumpfheimer** [reply]
I aggree. I kept the manufacturer specific configuration in mind. The exposure should be configured in zha quirks like [here](https://github.com/dumpfheimer/zha-device-handlers/blob/c14a9c6169352ab50f55ce49f110c65204895fb9/zhaquirks/xiaomi/aqara/opple_switch.py#L106) .

Just like the attributes it should be possible to add / remove / overwrite them.

**Comment by dmulcahey** [comment]
Once the ZCL 7/8 PR that was merged into Zigpy is in HA the goal is to support this type of thing w/o having to explicitly specify things the same way that z2m does ("exposes") otherwise every device would need to be quirked. The part of this that is in Zigpy may be handled in HA itself (or the web socket server) I started the beginning of this type of thing w/ sirens. We already expose several selects for things like alert tone, volume and strobe. The goal is to extend the support w/ number fields and additional selects, switches etc. tied directly to ZCL commands / attributes. Currently things look like this: https://github.com/home-assistant/core/blob/cdd23abea7f8854d95c22a2dadeaffdbceeccc55/homeassistant/components/zha/select.py#L47 Next we'll tie these to ZCL commands for things like power on state etc.

**Comment by dumpfheimer** [comment]
Thanks four your replies.

My first approach actually was to expose all the attributes but there were too many attributes and they were causing a lot of traffic.
I then chose this more permissive approach wich needs explicit declaration of the attributes but would make sure it actually makes sence to expose them.

I also thought about exposing attributes by default, if they have a class - wrapper with defined values. That could tackle most of them and should only be at least somewhat usefull attributes as someone actually took the time to write it.

@dmulcahey I was inspired by the class you tagged and it was the basis of my development. The biggest issue I had was that it only catches the attributes that have not been claimed before, which in my case was none. Also this approach gives the quirks the possibility to disable / overwrite the exposure defined in an underlying cluster definition.

@MattWestb startup leve, color and state are exposed by default. you will not be happy with the color-chooser, though. it just takes an integer, not a color.

![image](https://user-images.githubusercontent.com/60989727/158417719-25f7119d-d4af-4099-958c-43a1306f265c.png)

Just random thoughts now
Things I tried to keep in mind:

- Very little impact on zigpy
- No impact on zigpy's underlying controller-specific implementation
- Easy to understand configuration
- It is possible to configure exposure for all / very many devices in one spot
- It is possible to overwrite exposure for sepecific / few devices in quirks


Now the question is: could this become a viable solution or will another approach be preferred leaving no room for this one.

**Reply by dmulcahey** [reply]
> @dmulcahey I was inspired by the class you tagged and it was the basis of my development. The biggest issue I had was that it only catches the attributes that have not been claimed before, which in my case was none. Also this approach gives the quirks the possibility to disable / overwrite the exposure defined in an underlying cluster definition.

Did you use strict match or multi match? W/ multi match t the claiming doesn't cause any issues. for instance in the web socket server implementation I am working on I have several sensors linked to the basic cluster. You can actually see this in ZHA proper too w/ LQI / RSSI sensors. Anything currently tagged with strict match would need to change to multi match including the new additions.

My preference is to expose the most common functionalities (things like power on state, sensitivities, reset times, etc. Things that are used configurable w/ attributes writes or commands and are used by many users) via direct class definitions in ZHA / zha-websocket-server leveraging multi match so that we define them once and they just work for all devices (and for things that aren't spec compliant we can continue to intercept the calls and translate as necessary in quirks). That said I am open to talking through alternatives.

Personally, I don't think this belongs in Zigpy but I would be interested to get the opinions of @puddly and @Adminiuga. All of this is left up to the host application for now and we may not want to force these constraints across Zigpy users (HA, Domocitz, Jeedom, etc.) The way we handled device automation triggers in quirks was a complete hack (which I am responsible for) and I don't want to make the same mistake again but as stated above I welcome a discussion.

**Reply by dumpfheimer** [reply]
I tried the multi match which I was not able to get working. After thinking about how complex this would get for manufacturer specific things (adding for certain devices, removing existing ones for others) I thougt it might not be the purpose of the annotation.

Pretty much all of the code is in the ZHA integration, not zigpy. I just added the exposure there because the attributes were defined there too. I am open to moving them somewhere else.

with quirks I was wondering a few days ago if it was possible to replace/extend clusters on a manufacturer level or dynamically creating device automation triggers, do you happen to know?

**Reply by dmulcahey** [reply]
> with quirks I was wondering a few days ago if it was possible to replace/extend clusters on a manufacturer level or dynamically creating device automation triggers, do you happen to know?

I don't know if I understand what you mean?

**Reply by dmulcahey** [reply]
> I tried the multi match which I was not able to get working. After thinking about how complex this would get for manufacturer specific things (adding for certain devices, removing existing ones for others) I thougt it might not be the purpose of the annotation.

With multi match what you did in discovery: discover_configuration_entities and the wrapper functions around the class definitions  in the platform modules aren't necessary. what specifically wasn't working when you attempted to use it? As I mentioned above any existing strict match annotations for the clusters you tried would also have to be converted to multi match.

**Reply by dumpfheimer** [reply]
Its a seperate thought. Just wanted to ask you when I have you "on the line" already.

For example here: https://github.com/dumpfheimer/zha-device-handlers/blob/c14a9c6169352ab50f55ce49f110c65204895fb9/zhaquirks/xiaomi/aqara/opple_switch.py#L83

1. Is it possible to apply this cluster to all xiaomi devices that have a specific cluster id? (not arguing this specific case makes sence)
2. Is it possible to dynamically create device automation triggers? Xiaomi uses predictable cluster ids that send defined events. (eg 41 is button 1, 42 is button2, 51 is button1+button2, 61 would be 3 buttons .... not sure if these are the real cluster id's, but that is the idea behindit)

Both would imho greatly reduce code and code complexity and would support more devices out of the box.

**Reply by dmulcahey** [reply]
take a peek at this commit too: https://github.com/dmulcahey/home-assistant/commit/0a5dc65b13fdb7a70f653dfda3036a6dec776f9d I haven't PR'd this particular entity yet (honestly thought I had) but it contains a new number entity that writes to a ZCL attribute and you can see that it doesn't need any changes to discovery or any functions wrapped around class definitions.

**Reply by dmulcahey** [reply]
> Its a seperate thought. Just wanted to ask you when I have you "on the line" already.
> 
> For example here: https://github.com/dumpfheimer/zha-device-handlers/blob/c14a9c6169352ab50f55ce49f110c65204895fb9/zhaquirks/xiaomi/aqara/opple_switch.py#L83
> 
> 1. Is it possible to apply this cluster to all xiaomi devices that have a specific cluster id? (not arguing this specific case makes sence)
> 2. Is it possible to dynamically create device automation triggers? Xiaomi uses predictable cluster ids that send defined events. (eg 41 is button 1, 42 is button2, 51 is button1+button2, 61 would be 3 buttons .... not sure if these are the real cluster id's, but that is the idea behindit)
> 
> Both would imho greatly reduce code and code complexity and would support more devices out of the box.

Let's discuss that either on discord or in another discussion so we don't cross this thread up w/ things that are unrelated.

**Reply by dumpfheimer** [reply]
Would you create a new class for every attribute that shall be exposed as configurable?

**Reply by dmulcahey** [reply]
> Would you create a new class for every attribute that shall be exposed as configurable?

Take a look at the sensor classes. They would end up looking like that. Lots of small extensions that just specify the attribute.

**Reply by dumpfheimer** [reply]
Sorry for taking so long to respond. Will it somehow be possible to add these outside of ZHA (eg quirks)? What would the strategy be for adding manufacturer specific configuration attributes?

**Reply by dmulcahey** [reply]
Quirks for mfg specific attr and command defs

**Reply by dumpfheimer** [reply]
Okay, looks like you have this pretty much covered. May I ask, what is the state of development? Is there something I could help with?

Also, is there a board or page where I could have picked up that you are already working on this issue?
Thanks for your support.

**Reply by dmulcahey** [reply]
There isn’t a board (this is something we need to get better with) 

As for help, We’ll take all the help we can get. If you have time ping me on discord and we can chat a bit.

**Comment by Hedda** [comment]
> My preference is to expose the most common functionalities (things like power on state, sensitivities, reset times, etc. Things that are used configurable w/ attributes writes or commands and are used by many users) via direct class definitions in ZHA / zha-websocket-server leveraging multi match so that we define them once and they just work for all devices (and for things that aren't spec compliant we can continue to intercept the calls and translate as necessary in quirks). That said I am open to talking through alternatives.
> 
> **Personally, I don't think this belongs in Zigpy but I would be interested to get the opinions of @puddly and @Adminiuga. All of this is left up to the host application for now and we may not want to force these constraints across Zigpy users (HA, Domocitz, Jeedom, etc.)** The way we handled device automation triggers in quirks was a complete hack (which I am responsible for) and I don't want to make the same mistake again but as stated above I welcome a discussion.

 @pipiche38 & @zoic21 Perhaps this discussion is also of interest to Domoticz and Jeedom host application implementations using zigpy?

**Reply by pipiche38** [reply]
From a Domoticz standpoint , we are not at that level and not impacted. When I did the first test, I was requesting to get the Bind and Configure Reporting also done at the quirk/zigpy level, but got a clear push back.

**Reply by dmulcahey** [reply]
@pipiche38 take a look at: https://github.com/zigpy/zha-websocket-server timing just didn't line up when you folks were starting. this will provide a consistent application for anyone that wants to use it.

**Comment by Hedda** [comment]
Related common question/request for ZHA is configuration of StartUp/PowerOn values for Zigbee lights and switches from Philips Hue, see:

https://community.home-assistant.io/t/zha-hue-bulbs-restore-state-on-power-on/350548/

https://community.home-assistant.io/t/configure-ts0002-startup-behavior-and-switch-mode/410772

https://community.home-assistant.io/t/accessing-philips-hue-bulb-sensor-configuration-with-zha-integration/209638

**Comment by dumpfheimer** [comment]
@dmulcahey 
Would you take a look at my latest development?
I am aiming to expose the start up on/off and level configurations. This time I used the MULTI_MATCH annotation.
However, I had to change a crutial part in the code to get it to work: now in `discover_multi_entities` by default not only the remaining (unclaimed) channels are matched but also already claimed channels. [See here](https://github.com/dumpfheimer/core/blob/dd3d914798dbd153f6cac35beeb624e56931a587/homeassistant/components/zha/core/discovery.py#L186)

Branch:
https://github.com/dumpfheimer/core/tree/dumpfheimer/zha-start-up-on-off-select

**Comment by dmulcahey** [comment]
Getting closer. I was messing with this weekend too. Nit: Keep the number entity stuff separate to make this easier to review. 

The cache update from attribute writes needs to be done in Zigpy. I’ll open an issue for that. I need to look closer at this to determine what the impact to the claiming process is. I’ll pull this down and poke at it in the debugger in a bit. If you’re on discord we can chat about this.

**Reply by dumpfheimer** [reply]
How can I find you on discord?

**Reply by dmulcahey** [reply]
Same handle as here in the HomeAssistant server

**Comment by Hedda** [comment]
FYI, spotted new related question in HA community if ZHA exposes "switch_type" for Aqara/Xiaomi switch module T1 DLKZMK11LM or not?

https://community.home-assistant.io/t/aqara-switch-module-t1-rebound-switch/345580/

There is a related device support request asking for `switch_type` configuration of Aqara T1 switch module with neutral here:

https://github.com/zigpy/zha-device-handlers/issues/1231

I do not have that device but as I understand it, this "switch_type" configuration attribute allows you to change switch type from an ON/OFF rocker switch circuit-breaker type of switch to a NO (Normally Open) momentary push-button switch.

That "switch_type" is exposed in Zigbee2MQTT:

https://www.zigbee2mqtt.io/supported-devices/#e=switch_type

https://www.zigbee2mqtt.io/devices/DLKZMK11LM.html

# Xiaomi DLKZMK11LM

|     |     |
|-----|-----|
| Model | DLKZMK11LM  |
| Vendor  | Xiaomi  |
| Description | Aqara single switch module T1 (with neutral) |
| Exposes | switch (state), power, energy, temperature, voltage, power_outage_memory, led_disabled_night, **switch_type**, linkquality |
| Picture | ![Xiaomi DLKZMK11LM](https://www.zigbee2mqtt.io/images/devices/DLKZMK11LM.jpg) |

**Comment by dumpfheimer** [comment]
@dmulcahey has been working on exposing the configuration entities here: https://github.com/dmulcahey/home-assistant/tree/dm/zha-zcl-select

Also, I have a similar PR pending here: https://github.com/zigpy/zha-device-handlers/pull/1391
And I have added the necessary entites here: https://github.com/dumpfheimer/core/tree/dumpfheimer/zha-expose-configuration-entities

All three together make this setting available **for Aqara H1 Switches** but porting it for T1 _should_ not be very difficult.

**Reply by Hedda** [reply]
> FYI, spotted new related question in HA community if ZHA exposes "switch_type" for Aqara/Xiaomi switch module T1 DLKZMK11LM or not?
> 
> https://community.home-assistant.io/t/aqara-switch-module-t1-rebound-switch/345580/
> 
> There is a related device support request asking for `switch_type` configuration of Aqara T1 switch module with neutral here:
> 
> [zigpy/zha-device-handlers#1231](https://github.com/zigpy/zha-device-handlers/issues/1231)
> 
> I do not have that device but as I understand it, this "switch_type" configuration attribute allows you to change switch type from an ON/OFF rocker switch circuit-breaker type of switch to a NO (Normally Open) momentary push-button switch.

For reference, there is a similar request for "switch_type" configuration of Tuya TS0003 3 gang switch module (TZ3000_odzoiovu) here:

https://github.com/zigpy/zha-device-handlers/issues/1145

**Reply by dumpfheimer** [reply]
I have already implemented the switch type for the H1 you can use it as a reference (not merged yet ) https://github.com/zigpy/zha-device-handlers/pull/1391

**Reply by Ulrar** [reply]
Just want to mention that it seems to be only for the H1 with neutral, the H1 without neutral has a different signature for some reason and doesn't work with the handler from that PR.
See https://github.com/zigpy/zha-device-handlers/issues/1693

**Comment by Hedda** [comment]
FYI, tilt values window covering (blinds) device is another one needs exposing as reported by @cremor in https://github.com/zigpy/zha-device-handlers/discussions/1430

> I bought a window covering (blinds) device that doesn't show tilt values in HA. But I can use the Zigbee attribute "current_position_tilt_percentage" (0x0009) in the cluster "WindowCovering" (0x0102 in endpoint 6) to get the tilt value. So the device seems to report it correctly, but ZHA isn't using it.
> 
> How would a quirk for that need to look like? I assume overriding the clusters with the exact same ones that are there originally wouldn't change anything?
> 
> Maybe this is related to the problem that it looks like the device is identified as a window? At least it's shown with a window icon. If that is the case, how would I tell ZHA that this is a blind - so that it supports tilt?
> 
> It's about this device support request: https://github.com/zigpy/zha-device-handlers/issues/1397

* [Device Support Request] Insta GmbH Nexentro Blinds Actuator Mini -> https://github.com/zigpy/zha-device-handlers/issues/1397

* [Device Support Request] NHPB/SHUTTER/1 by Schneider Electric -> https://github.com/zigpy/zha-device-handlers/issues/1685

* [Device Support Request] TS130F - _TZ3000_1dd0d5yi - Moes curtain switch -> https://github.com/zigpy/zha-device-handlers/issues/1416

* [Device Support Request] Smartwings WM25/L-Z Blinds -> https://github.com/zigpy/zha-device-handlers/issues/1689

* [Device Support Request] Tuya Curtain Motor -> https://github.com/zigpy/zha-device-handlers/issues/744

**Reply by cremor** [reply]
While it would be nice to get those tilt values shown in the UI, I think a better solution would be to natively support tilt for covers in ZHA. Those Zigbee attributes are the default ones defined by the Zigbee standard, so HA/ZHA should support those and show them in the default cover UI elements, just like it does for Z-Wave.

**Reply by Hedda** [reply]
Yes for comparison see some of the other components in Home Assistant core which already implemented tilt, like zwave-js and mqtt:

https://github.com/home-assistant/core/blob/0e674fc59746fb648cc7c3b736e87c2abeb69acd/homeassistant/components/zwave_js/cover.py

https://github.com/home-assistant/core/blob/817f0c9aae2b590a069dbc990a90b278e490e065/tests/components/zwave_js/test_cover.py

https://github.com/home-assistant/core/blob/fe38e6ba875cb267da2ed066109b946a124166cc/homeassistant/components/mqtt/cover.py

https://github.com/home-assistant/core/blob/fe38e6ba875cb267da2ed066109b946a124166cc/tests/components/mqtt/test_cover.py

https://github.com/home-assistant/core/blob/9b5939a7a5424a8545c78e9b6b5f20e1a7ab7ac7/homeassistant/components/demo/cover.py

https://github.com/home-assistant/core/blob/e9a9b119c40c9337b1c61b7780e97b689cbd9f0f/homeassistant/components/template/cover.py

https://github.com/home-assistant/core/blob/3f7b7187ab410fd8dbf04fcfc834fda731dc8087/tests/components/group/test_cover.py

https://github.com/home-assistant/core/blob/a59460a23336627d0bc12b1eefffdaa516e55e87/tests/components/cover/test_reproduce_state.py

Some other components that also already include tilt support for cover are; deconz, somfy, esphome, knx, rfxtrx, and many more:

https://github.com/home-assistant/core/search?p=4&q=tilt

**Comment by TheKingOfAtlantis** [comment]
Hey, coming in as a bit of an outsider but if any further development has taken place to expose device configs, I was recently trying to change the speed on a aqara blind roller with HA but instead had to find right cluster to adjust it, so was curious what might happen with regards to this?

**Comment by dumpfheimer** [comment]
Configuration entities have been decided to be a zha/application specific thing only, which means they by design do not fit into the quirks. I think you will have to create a PR in Home Assistant to add a configuration entity.

**Comment by Hedda** [comment]
FYI, there were some mentionings related to this in the latest "Building the Open Home" article as well as in "State of the Open Home 2022":

https://building.open-home.io/innovation-versus-standards-in-the-open-home/ ("How a pet feeder shows that standards don't solve everything").

https://www.youtube.com/watch?v=D936T1Ze8-4&ab_channel=HomeAssistant (Nabu Casa user experience and frontend developers for example discussed their goal of making Home Assistant UI become more device-oriented in the future as opposed to being entities-oriented).

> Configuration entities have been decided to be a zha/application specific thing only, which means they by design do not fit into the quirks. I think you will have to create a PR in Home Assistant to add a configuration entity.

Meaning if wish to expose additional device configuration attributes by default then extending the device types in zha component is needed:

https://github.com/home-assistant/core/tree/dev/homeassistant/components/zha

i.e. submitting pull requests if you are a Python developer: 

https://github.com/home-assistant/core/pulls?q=is%3Apr+zha

As per existing pull requests as examples:

https://github.com/home-assistant/core/pull/81520 (Add TI router transmit power config entity to ZHA )
https://github.com/home-assistant/core/pull/82340 (Add Aqara c1 pet feeder support to ZHA)
https://github.com/home-assistant/core/pull/80853 (Add ZHA StartUpColorTemperature configuration entity)
https://github.com/home-assistant/core/pull/79923 (Add Philips Hue motion sensor config entities to ZHA)
https://github.com/home-assistant/core/pull/79106 (Add configuration entities and device actions for Inovelli Blue Series switch to ZHA)
https://github.com/home-assistant/core/pull/82087 (Add config entities for lumi.motion.agl04)
https://github.com/home-assistant/core/pull/82596 (Add GiEX valve support to ZHA)
https://github.com/home-assistant/core/pull/81218 (Add Tuya Backlight mode configuration)

and end users will need to start a new device support request(s):

https://github.com/zigpy/zha-device-handlers/issues/new/choose

and maybe also open a new topic under feature requests in the Home Assistant forum about exposing specific device(s) configurations(?):

https://community.home-assistant.io/c/feature-requests/13
